pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,489 @@
+		match message.get_message_proof() {
+			None => message
+				.to_message_proof()
+				.get_message_proof()
+				.expect("message proof ensured by 'to_message_proof'"),
+			Some(proof) => proof,
+		}
+	}
+
+	/// Creates an inbound entry based on whether the inbound message is a
+	/// proof or not.
+	fn create_inbound_entry(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		message: T::Message,
+	) -> InboundEntry<T> {
+		match message.get_message_proof() {
+			None => InboundEntry::Message {
+				session_id: inbound_processing_info.current_session_id.clone(),
+				domain_address: inbound_processing_info.domain_address.clone(),
+				message,
+				expected_proof_count: inbound_processing_info.expected_proof_count_per_message,
+			},
+			Some(_) => InboundEntry::Proof {
+				session_id: inbound_processing_info.current_session_id,
+				current_count: 1,
+			},
+		}
+	}
+
+	/// Validation ensures that:
+	///
+	/// - the router that sent the inbound message is a valid router for the
+	///   specific domain.
+	/// - messages are only sent by the first inbound router.
+	/// - proofs are not sent by the first inbound router.
+	fn validate_inbound_entry(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		router_id: &T::RouterId,
+		inbound_entry: &InboundEntry<T>,
+	) -> DispatchResult {
+		let router_ids = inbound_processing_info.router_ids.clone();
+
+		ensure!(
+			router_ids.iter().any(|x| x == router_id),
+			Error::<T>::UnknownRouter
+		);
+
+		match inbound_entry {
+			InboundEntry::Message { .. } => {
+				ensure!(
+					router_ids.get(0) == Some(&router_id),
+					Error::<T>::MessageExpectedFromFirstRouter
+				);
+
+				Ok(())
+			}
+			InboundEntry::Proof { .. } => {
+				ensure!(
+					router_ids.get(0) != Some(&router_id),
+					Error::<T>::ProofNotExpectedFromFirstRouter
+				);
+
+				Ok(())
+			}
+		}
+	}
+
+	/// Upserts an inbound entry for a particular message, increasing the
+	/// relevant counts accordingly.
+	fn upsert_pending_entry(
+		message_proof: Proof,
+		router_id: T::RouterId,
+		inbound_entry: InboundEntry<T>,
+		weight: &mut Weight,
+	) -> DispatchResult {
+		weight.saturating_accrue(T::DbWeight::get().writes(1));
+
+		PendingInboundEntries::<T>::try_mutate(message_proof, router_id, |storage_entry| {
+			match storage_entry {
+				None => {
+					*storage_entry = Some(inbound_entry);
+
+					Ok::<(), DispatchError>(())
+				}
+				Some(stored_inbound_entry) => match stored_inbound_entry {
+					InboundEntry::Message {
+						session_id: stored_session_id,
+						expected_proof_count: stored_expected_proof_count,
+						..
+					} => match inbound_entry {
+						InboundEntry::Message {
+							session_id: new_session_id,
+							expected_proof_count: new_expected_proof_count,
+							..
+						} => {
+							if *stored_session_id != new_session_id {
+								*stored_session_id = new_session_id;
+								*stored_expected_proof_count = new_expected_proof_count;
```

Comment:
Q: Shouldn't we reset the proof count if the session id does not match? I cannot tell if this always happens here, i.e. can `new_expected_proof_count` be at most 1?

Update: Ah I see that happens here: https://github.com/centrifuge/centrifuge-chain/pull/1961/files#diff-82099ad70b6f867f15a4b001aa9cfad2a5a8bf0a84cc133ec1d46f5bf78d3f26R314

Will still post this comment to be sure we are all alligned.

Reply:
Yes, will change this to be explicitly 1 not to leave any door open for any bugs in the future.

Reply:
Nvm, this is correct. If the session changed, we take the new count from the entry, if not, we add it to the existing count.

Reply:
A safer and clearer way would be to just overwrite the entry if the session changed.